File: app/src/main/java/com/dmi/perfectreader/render/RenderObject.java
package com.dmi.perfectreader.render;

import android.graphics.Canvas;

import java.util.List;

public abstract class RenderObject {
    private final float width;
    private final float height;
    private final List<RenderChild> children;

    public RenderObject(float width, float height, List<RenderChild> children) {
        this.width = width;
        this.height = height;
        this.children = children;
    }

    public float width() {
        return width;
    }

    public float height() {
        return height;
    }

    public List<RenderChild> children() {
        return children;
    }

    public RenderChild child(int index) {
        return children.get(index);
    }

    /**
     * Необходимо для механизма разделения на страницы.
     *   true - для текстовых строк, изображений, и всех тех объектов, которые необходимо рисовать на экране целиком
     *   false - для контейнеров, таблиц, и всех тех объектов, которые можно рисовать частично (обрезав, например верхнюю границу)
     */
    public abstract boolean canPartiallyPainted();

    /**
     * Нарисовать только содержимое данного объекта, без рисования дочерних объектов
     */
    public void paintItself(Canvas canvas) {
    }

    public void paintRecursive(Canvas canvas) {
        paintItself(canvas);
        // не использовать for-each! снижается производительность
        for (int i = 0; i < children.size(); i++) {
            RenderChild child = children.get(i);
            canvas.translate(child.x(), child.y());
            child.object().paintRecursive(canvas);
            canvas.translate(-child.x(), -child.y());
        }
    }

    public boolean click(float x, float y) {
        return false;
    }
}

File: app/src/main/kotlin/com/dmi/perfectreader/render/RenderObject.kt
package com.dmi.perfectreader.render

import android.graphics.Canvas

abstract class RenderObject(
        val width: Float,
        val height: Float,
        val children: List<RenderChild>
) {
    fun child(index: Int) = children[index]

    /**
     * Необходимо для механизма разделения на страницы.
     * true - для текстовых строк, изображений, и всех тех объектов, которые необходимо рисовать на экране целиком
     * false - для контейнеров, таблиц, и всех тех объектов, которые можно рисовать частично (обрезав, например верхнюю границу)
     */
    open fun canPartiallyPainted() = false

    /**
     * Необходимо для механизма разделения на страницы.
     * Ненулевые margins означают, что, если элемент находит сверху или снизу страницы, то можно обрезать margins этого элемента.
     * К примеру, если у элемента верхняя граница 16, то можно расположить элемент с позицией y = -16
     */
    open fun internalMargins() = Margins.Companion.ZERO

    /**
     * Нарисовать только содержимое данного объекта, без рисования дочерних объектов
     */
    open fun paintItself(canvas: Canvas) = Unit

    fun paintRecursive(canvas: Canvas) {
        paintItself(canvas)
        for (i in 0..children.size - 1) {
            with (children[i]) {
                canvas.translate(x, y)
                obj.paintRecursive(canvas)
                canvas.translate(-x, -y)
            }
        }
    }

    class Margins(val left: Float, val right: Float, val top: Float, val bottom: Float) {
        companion object {
            val ZERO = Margins(0F, 0F, 0F, 0F)
        }
    }
}
